kevm_pyk/src/kevm_pyk/solc_to_k.py

```diff
@@ -37,6 +39,198 @@
 _LOGGER: Final = logging.getLogger(__name__)
 
 
+@dataclass
+class Contract():
+
+    @dataclass
+    class Method:
+        name: str
+        id: int
+        sort: KSort
+        arg_names: List[str]
+        arg_types: List[str]
+        contract_name: str
+
+        def __init__(self, name: str, id: int, abi: Dict, contract_name: str, sort: KSort) -> None:
+            self.name = name
+            self.id = id
+            self.arg_names = [f'V{i}_{input["name"].replace("-", "_")}' for i, input in enumerate(abi['inputs'])]
+            self.arg_types = [input['type'] for input in abi['inputs']]
+            self.contract_name = contract_name
+            self.sort = sort
+
+        @property
+        def selector_alias_rule(self) -> KRule:
+            return KRule(KRewrite(KEVM.abi_selector(self.name), intToken(self.id)))
+
+        @property
+        def production(self) -> KProduction:
+            input_nonterminals = (KNonTerminal(_evm_base_sort(input_type)) for input_type in self.arg_types)
+            items_before: List[KProductionItem] = [KTerminal(self.name), KTerminal('(')]
+            items_args: List[KProductionItem] = list(intersperse(input_nonterminals, KTerminal(',')))
+            items_after: List[KProductionItem] = [KTerminal(')')]
+            return KProduction(self.sort, items_before + items_args + items_after, klabel=KLabel(f'method_{self.contract_name}_{self.name}'))
+
+        def rule(self, contract: KInner, application_label: KLabel, contract_name: str) -> KRule:
+            arg_vars = [KVariable(aname) for aname in self.arg_names]
+            prod_label = self.production.klabel
+            assert prod_label is not None
+            lhs = KApply(application_label, [contract, KApply(prod_label, arg_vars)])
+            args: List[KInner] = [KEVM.abi_type(input_type, KVariable(input_name)) for input_type, input_name in zip(self.arg_types, self.arg_names)]
+            rhs = KEVM.abi_calldata(self.name, args)
+            opt_conjuncts = [_range_predicate(KVariable(input_name), input_type) for input_name, input_type in zip(self.arg_names, self.arg_types)]
+            conjuncts = [opt_conjunct for opt_conjunct in opt_conjuncts if opt_conjunct is not None]
+            if len(conjuncts) == 0:
+                ensures = TRUE
+            else:
+                ensures = functools.reduce(lambda x, y: KApply('_andBool_', [x, y]), conjuncts)
+            return KRule(KRewrite(lhs, rhs), ensures=ensures)
+
+    @dataclass
+    class Field:
+        name: str
+        slot: int
+        type: str
+        sort: KSort
+        klabel: KLabel
+        types: Dict
+        _base_types: Set[str]
+
+        def __init__(self, name: str, slot: int, type: str, contract_name: str, field_sort: KSort, types: Dict) -> None:
+            self._base_types = {'t_address', 't_bool', 't_bytes4', 't_bytes32', 't_uint256', 't_int256', 't_uint8', 't_string_storage'}
```

Comment:
```suggestion
    @dataclass
    class Field:
        _BASE_TYPES: Final = {'t_address', 't_bool', 't_bytes4', 't_bytes32', 't_uint256', 't_int256', 't_uint8', 't_string_storage'}

        name: str
        slot: int
        type: str
        sort: KSort
        klabel: KLabel
        types: FrozenDict

        def __init__(self, name: str, slot: int, type: str, contract_name: str, field_sort: KSort, types: Dict) -> None:
```

To use `FrozenDict`:

```python3
from pyk.utils import FrozenDict
```

Reply:
Done.